### PR TITLE
Rollback fee

### PR DESCRIPTION
**Pull request by dolgopolovwork**
Now we can rollback fee.

### CODE REVIEW

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/handler/BroadcastTransactionHandler.kt [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        val withdrawalDetails = gson.fromJson(broadcastCommand.value.irohaUnEscape(), WithdrawalDetails::class.java)
+        feeService.payFee(withdrawalDetails)
+            .fold(
+                { logger.info("Fee for withdrawal $withdrawalDetails has been payed") },
+                { ex -> logger.error("Cannot pay fee for withdrawal $withdrawalDetails", ex) }
+            )
+    }
+
+    companion object : KLogging()
+}
```
Comment (Alexey-N-Chernyshov) [inline btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/handler/BroadcastTransactionHandler.kt]: 😱

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/service/FeeService.kt [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import com.github.kittinunf.result.Result
+import com.github.kittinunf.result.flatMap
+import com.github.kittinunf.result.map
+import org.springframework.beans.factory.annotation.Qualifier
+import org.springframework.stereotype.Component
+import java.math.RoundingMode
+
+private const val BTC_ASSET_ID = "btc#bitcoin"
+// Key that signals that transfer is 'fee'
+const val FEE_KEY = "f842c3e1858df8c85794efacd06871f806afab04e0458c44488d2ffb3f071075"
```
Comment (Alexey-N-Chernyshov) [inline btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/service/FeeService.kt]: I suggest to move fee description to notary repository.

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/handler/BroadcastTransactionHandler.kt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import iroha.protocol.Commands
+import mu.KLogging
+import org.springframework.stereotype.Component
+
+/**
+ * Handler that handles 'broadcast' events
+ */
+@Component
+class BroadcastTransactionHandler
+    (private val feeService: FeeService) {
```
Comment (Mingela) [inline btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/handler/BroadcastTransactionHandler.kt]: strange formatting

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/handler/BroadcastTransactionHandler.kt (hunk 1 above)
Comment (Mingela) [inline btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/handler/BroadcastTransactionHandler.kt]: add a line please

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/BroadcastsProvider.kt [hunk 4]
```diff
@@ -21,6 +23,7 @@ class BroadcastsProvider(
     private val withdrawalQueryHelper: IrohaQueryHelper
 ) {
 
+    private val gson = Gson()
```
Comment (Mingela) [inline btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/BroadcastsProvider.kt]: how about using a separate `static` `Gson` somewhere from utility class in all the classes need this

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/service/FeeService.kt [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        val domain = withdrawalDetails.sourceAccountId.split("@")[1]
+        return withdrawalConsumer.getConsumerQuorum()
+            .flatMap { quorum ->
+                ModelUtil.transferAssetIroha(
+                    withdrawalConsumer,
+                    withdrawalConsumer.creator,
+                    "withdrawal_billing@$domain",
+                    BTC_ASSET_ID,
+                    "Fee",
+                    satToBtc(withdrawalDetails.withdrawalFeeSat).setScale(8, RoundingMode.DOWN).toPlainString(),
```
Comment (Mingela) [inline btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/service/FeeService.kt]: `RoundingMode.UP` ?

**Review by Mingela — APPROVED** [review]

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/service/FeeService.kt [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+     * @return result of operation
+     */
+    fun payFee(withdrawalDetails: WithdrawalDetails): Result<Unit, Exception> {
+        val domain = withdrawalDetails.sourceAccountId.split("@")[1]
+        return withdrawalConsumer.getConsumerQuorum()
+            .flatMap { quorum ->
+                ModelUtil.transferAssetIroha(
+                    withdrawalConsumer,
+                    withdrawalConsumer.creator,
+                    "withdrawal_billing@$domain",
```
Comment (Alexey-N-Chernyshov) [inline btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/service/FeeService.kt]: move to config

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/service/FeeService.kt (hunk 5 above)
Comment (dolgopolovwork) [reply btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/service/FeeService.kt]: I think in this case user may be out of fee.

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/BroadcastsProvider.kt (hunk 4 above)
Comment (dolgopolovwork) [reply btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/provider/BroadcastsProvider.kt]: Good idea

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/handler/NewTransferHandler.kt [hunk 7]
```diff
@@ -35,25 +35,33 @@ class NewTransferHandler(
...
             "Withdrawal event(" +
                     "from:$sourceAccountId " +
                     "to:$destinationAddress " +
-                    "amount:${btcAmount.toPlainString()}" +
+                    "amount:${btcAmount.toPlainString()} " +
+                    "fee:$fee " +
                     "hash:${withdrawalDetails.irohaFriendlyHashCode()})"
         }
-
+        //TODO check withdrawal before this call
```
Comment (Alexey-N-Chernyshov) [inline btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/handler/NewTransferHandler.kt:64]: What does it actually mean?

File: btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/init/BtcWithdrawalInitialization.kt [hunk 8]
```diff
@@ -186,6 +196,36 @@ class BtcWithdrawalInitialization(
...
+        var withdrawalCommand: Commands.Command? = null
+        var feeCommand: Commands.Command? = null
+        transaction.payload.reducedPayload.commandsList.forEach { command ->
+            if (command.hasTransferAsset()) {
+                val transferCommand = command.transferAsset
+                if (isValidBtcAddress(transferCommand.description)) {
+                    // If description is a valid BTC address, then it's a withdrawal command
+                    withdrawalCommand = command
+                } else if (transferCommand.description == FEE_DESCRIPTION) {
+                    // If description is a sha256('Fee'), then it's a fee command
```
Comment (Alexey-N-Chernyshov) [inline btc-withdrawal/src/main/kotlin/com/d3/btc/withdrawal/init/BtcWithdrawalInitialization.kt]: comment is outdated

File: btc/src/main/kotlin/com/d3/btc/helper/format/FormatHelper.kt [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+import com.google.gson.Gson
+
+/**
+ * Gson object holder
+ */
+object GsonInstance {
+    private val gson = Gson()
+
+    fun get() = gson
+}
```
Comment (Alexey-N-Chernyshov) [inline btc/src/main/kotlin/com/d3/btc/helper/format/FormatHelper.kt]: 😱

**Review by Alexey-N-Chernyshov — APPROVED** [review]

**Review by Alexey-N-Chernyshov — APPROVED** [review]

**Pull request merged**